lib/matplotlib/widgets.py

```diff
@@ -1072,9 +1071,18 @@ def set_active(self, index):
         if index not in range(len(self.labels)):
             raise ValueError(f'Invalid CheckButton index: {index}')
 
-        l1, l2 = self.lines[index]
-        l1.set_visible(not l1.get_visible())
-        l2.set_visible(not l2.get_visible())
+        if colors.same_color(
+                self._crosses.get_facecolor()[index], colors.to_rgba("none")
+        ):
+            self._crosses.get_facecolor()[index] = colors.to_rgba("k")
+        else:
+            self._crosses.get_facecolor()[index] = colors.to_rgba("none")
+
+        if hasattr(self, "_rectangles"):
+            for i, p in enumerate(self._rectangles):
+                p.set_facecolor("k" if colors.same_color(
+                    p.get_facecolor(), colors.to_rgba("none"))
+                                else "none")
 
         if self.drawon:
             self.ax.figure.canvas.draw()
```

Comment:
@anntzer I have a preliminary implementation for square check boxes but I'm running into one small issue. When I try running this, everything works as intended when we change the figure size (the check boxes remain square), but the unit test fails. For some reason, calling `set_active(0)` in code doesn't change the status of the checkboxes. The issue is arising in these lines. The status changes correctly based on the code that I added but after executing `self.ax.figure.canvas.draw()`, the colors for all the scatter markers change back to what they were initially set to be, and not the changed ones. This is happening only in the unit tests and not while actually checking it using a GUI backend. Any ideas/suggestions?

Reply:
I feel like I'm missing something but can't see it right now.

Reply:
If you do some print-debugging and grep for places that could touch facecolors, you'll see that Collection.update_scalarmappable gets called at some point which causes the facecolors to be reset, but you'll need to figure out by yourself why (because I don't actually know immediately).

Reply:
That helps, thanks! I'll try figuring out why this is happening.

Reply:
@anntzer this is actually true for the radio buttons too! Small code to reproduce (based on the checkbox example)

```python
radio = RadioButtons(rax, labels, active=1, activecolor="k")
print(radio.value_selected)
print(radio._buttons.get_facecolor())
print("Changing active button from 1 to 0")
radio.set_active(0)
print(radio.value_selected)
print(radio._buttons.get_facecolor())
```
Output:
```
4 Hz
[[0. 0. 0. 0.]
 [0. 0. 0. 1.]
 [0. 0. 0. 0.]]
Changing active button from 1 to 0
2 Hz
[[0. 0. 0. 0.]
 [0. 0. 0. 1.]
 [0. 0. 0. 0.]]
```

This shows that the facecolors for the buttons don't change even after `set_active(0)` is called, even though `value_selected` is correct. The problem arises because of these lines

https://github.com/matplotlib/matplotlib/blob/a9c96200f7c3eac1923310086df96072813bac7d/lib/matplotlib/collections.py#L842-L845

Since it sets the flags such that it returns True the first time, 

https://github.com/matplotlib/matplotlib/blob/a9c96200f7c3eac1923310086df96072813bac7d/lib/matplotlib/collections.py#L894-L900

set the original colors again. Since `update_scalarmappable` gets called the first time they're drawn, when creating a figure, all functionalities work as everything works as expected after the second call. Even in this case, if there are any `set_active` calls after the first one, they work fine.

Reply:
Thanks for the investigation, I have opened https://github.com/matplotlib/matplotlib/issues/24479 to track this.  Sorry for sending you down this rabbit hole, perhaps this was not so easy after all...

Reply:
Thanks, it was fun. But now with the question on hand, this bug stops me to run `get_status` for the check boxes correctly. I guess one workaround could be that I maintain an active array similar to `value_active` in radio buttons but that'll be a hack and I'm on the fence about having an extra array to handle that when it can be done with the colors itself. That'll pass the tests though, and everything works in a figure anyway. Suggestions?